js/error-handling.md

```diff
@@ -25,80 +25,154 @@ BAD
-// OUTPUT => "user { ... } already exists."
+new Promise(function(resolve, reject) {
+  throw new Error('Uh-oh!');
+}).catch(function(error) {
+  console.error(error);
+});
+
+// => Error: Uh-oh!
+//     at <anonymous>:2:9
+//     at new Promise (<anonymous>)
+//     at <anonymous>:1:16
 ```
 
 BAD
 
 ```js
-console.error(user + " already exists.");
-// OUTPUT => "[object Object] already exists."
+new Promise(function(resolve, reject) {
+  throw 'Uh-oh!';
+}).catch(function(error) {
+  console.error(error);
+});
+
+// => Error: Uh-oh!
 ```
 
-#### Define reusable error classes, provide details as error message
+#### Do not serialize error classes
 
 GOOD
 
 ```js
-class RecordAlreadyExistsError extends Error {
-  // ...
-}
+new Promise(function(resolve, reject) {
+  throw new Error('Uh-oh!');
+}).catch(function(error) {
+  console.error(error);
+});
 
-if (userExists(user)) {
-  throw new RecordAlreadyExistsError(JSON.stringify(user) + " already exists.");
-}
+// => Error: Uh-oh!
+//      at <anonymous>:2:44
+//      at new Promise (<anonymous>)
+//      at <anonymous>:2:1
+//      ...
 ```
 
 BAD
 
 ```js
-class UserError extends Error {
-  // ...
-}
+new Promise(function(resolve, reject) {
+  throw new Error('Uh-oh!');
+}).catch(function(error) {
+  console.error(JSON.stringify(error));
+});
 
-if (userExists(user)) {
-  throw new UserError("Already exists.");
-}
+// => {}
 ```
 
-#### Rejected Promises must always be catched and at least logged
+BAD
 
-Otherwise information might be lost.
+```js
+new Promise(function(resolve, reject) {
+  throw new ExampleError('Uh-oh!');
+}).catch(function(error) {
+  console.error(JSON.stringify(error));
+});
+
+// => {"name":"ExampleError"}
+```
+
+#### Serialize objects when wrapping into error classes
+
+Otherwise the message cannot be retreived.
 
 GOOD
 
 ```js
 new Promise(function(resolve, reject) {
-  throw new Error('Uh-oh!');
+  throw new Error(JSON.stringify({
+    message: 'Uh-oh!'
+  }));
 }).catch(function(error) {
   console.error(error);
-});
 
-// => Error: Uh-oh!
-//     at <anonymous>:2:9
-//     at new Promise (<anonymous>)
-//     at <anonymous>:1:16
+  // => Error: {"message":"Uh-oh!"}
+  //      at <anonymous>:68:50
+  //      at new Promise (<anonymous>)
+  //      ...
+
+  console.error("Uh! " + error);
+
+  // => Uh! Error: {"message":"Uh-oh!"}
+});
 ```
 
 BAD
 
 ```js
 new Promise(function(resolve, reject) {
-  throw new Error('Uh-oh!');
+  throw new Error({
+    message: 'Uh-oh!'
+  });
+}).catch(function(error) {
+  console.error(error);
+
+  // => Error: [object Object]
+  //      at <anonymous>:58:50
+  //      at new Promise (<anonymous>)
+  //      ...
+
+  console.error("Uh! " + error);
+
+  // => Uh! Error: [object Object]
 });
+```
 
-// => (node:12977) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): Error: Uh-oh!
+#### Define reusable error classes, provide details as error message
```

Comment:
```suggestion
#### Define generic purpose error classes, provide details as error message
```

I know this wasn't modified now but imo it was not obvious why one is better than the other. Maybe this phrasing helps. Or do you have any proposals? Is it clear for you?

Reply:
IMO both headings are meaningful enough, perhaps the one you've just proposed is slightly better.

Going to add this change.